MSC3827: Filtering of `/publicRooms` by room type (#3827)
matrix-org/matrix-spec-proposals: proposals/3827-space-explore.md

# MSC3827: Filtering of `/publicRooms` by room type

[MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) defines Spaces as
rooms with `type` of `m.space`. One of the most requested features for Spaces is
to have [a way to to discover
them](https://github.com/vector-im/element-web/issues/17264). To solve this
issue, this MSC proposes a mechanism for filtering the response of
`/publicRooms` by room type.

## Proposal

This MSC proposes adding a `room_type` filter to the `/publicRooms` endpoint.
The value of `room_type` is an array of room types - possible values of the

> `type` field in the `m.room.create` state event - which should be returned.

Reviewer: Does this make it possible to filter for non-spaces, e.g. rooms without a `room_type`? I guess the following could do this:

```json
{
    "filter": {
        "room_type": []
    }
}
```

But this precludes searching for spaces and "normal" rooms while ignoring other types of rooms?

Author: Hmm, the ideal would be adding a default room type (`m.room`) - is that planned, has it ever been discussed, would it ok using `m.room` just in this MSC?

Reviewer: Another option could be to use a special `null` to mean "missing"? Not pretty but eh.

Author: > Another option could be to use a special `null` to mean "missing"? Not pretty but eh.

Yeah, that's not exactly the nicest but would work...

An example request could look like this:

```HTTP
POST /_matrix/client/v3/publicRooms
{
    "filter": {
        "room_type": ["m.space"]
    },
    "limit": 10
}
```

> the response then also includes a `room_type` field, so that clients can show

Reviewer: I assumed the intention was to always show it, but it doesn't quite say that.

```suggestion
The response is also modified to include a `room_type` field, so that clients can show
```

> the response then also includes a `room_type` field, so that clients can show

Author: Would `servers should now also include a `room_type` field in the response, so that clients can show the room type metadata.` sound more clear?

> the response then also includes a `room_type` field, so that clients can show

Author: > I assumed the intention was to always show it, but it doesn't quite say that.

This is probably the clearest :+1:

> the room type metadata.

Reviewer: 🎉  This brings it closer in-line to `/hierarchy`!

# MSC3827: Filtering of `/publicRooms` by room type

[MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) defines Spaces as
rooms with `type` of `m.space`. One of the most requested features for Spaces is
to have [a way to to discover
them](https://github.com/vector-im/element-web/issues/17264). To solve this
issue, this MSC proposes a mechanism for filtering the response of
`/publicRooms` by room type.

## Proposal

> This MSC proposes adding a `room_types` filter to the `/publicRooms` endpoint.

Reviewer 2: Nitpick: since we have both `POST` and `GET` versions of the endpoint, it should specify which one it applies to

```suggestion
This MSC proposes adding a `room_types` filter to the `POST /publicRooms` endpoint.
```

(presumably, this is only being added to the `POST` version, since the `GET` version doesn't have a filter).

> This MSC proposes adding a `room_types` filter to the `POST /publicRooms` endpoint.

Reviewer 3: please link to the original spec, to help readers of the MSC.

```suggestion
This MSC proposes adding a `room_types` filter to the [`POST /publicRooms`](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3publicrooms) endpoint.
```

The value of `room_types` is an array of room types - possible values of the

> `type` field in the `m.room.create` state event - which should be returned.

Reviewer 4: ```suggestion
`type` field in the `m.room.create` state event's `content` dictionary - which should be returned.
```

An example request could look like this:

```HTTP
POST /_matrix/client/v3/publicRooms
{
    "filter": {
        "room_types": ["m.space"]
    },
    "limit": 10
}
```

> The response is also modified to include a `room_type` field, so that clients can show

Reviewer 2: Again, we should specify whether the change applies to both `POST` and `GET`, or to only one.  Probably makes sense to change both (and I think that's what the synapse impl does), so

```suggestion
The response for both `POST /publicRooms` and `GET /publicRooms` is also modified to include a `room_type` field, so that clients can show
```
